torch/jit.py

```diff
@@ -1,26 +1,375 @@
 import torch.autograd.function as function
 import torch._C
 from torch.autograd import Variable
-from torch.nn import Module
+from torch.nn import Module, ParameterList, Parameter
 from collections import defaultdict
 import itertools
 import types
 import contextlib
 import os
+import functools
 import torch.contrib._graph_vis as graph_vis
-# Example how to use:
-#
-# import torch.jit
-# model = model.RNNModel(args.model, ...)
-# model = torch.jit.traced(model)
 
 
-def flatten(x):
+HOLE = object()
+VOLATILE = object()
+
+
+# TODO: verify is not implemented yet
+def compile(arg=None, nderivs=1, params=tuple(), verify=False, optimize=True, enabled=True):
+    """
+    Mark a function or module as eligible for just-in-time compilation.  The
+    next time the function/module is executed, it is traced, and the trace is
+    compiled into an optimized representation which is run in lieu of the
+    original Python code upon subsequent invocations of the function/module.
+
+    The result of this function is stateful, so make sure you invoke it
+    only once per code you want to JIT compile.
+
+    .. note::
+
+        If your function/module takes non-Variable inputs, the JIT compiler
+        will compile a trace separately for each distinct input configuration.
+
+    .. warning::
+
+        Just-in-time compilation currently only works for functions/modules
+        which do not have dynamic control flow; if you compile a function/module
+        which has dynamic control flow, you will silently get incorrect
+        results on subsequent invocations of the model.  Use `verify=True` to
+        check that the original Python code and optimized code are equivalent.
```

Comment:
This warning is unclear to me. What's "dynamic control flow"? E.g. having a loop that depends on an int argument to the function is fine.

Reply:
New copy:

```

        Just-in-time compilation currently only works for functions/modules
        which are not data dependent (e.g., have conditionals on data in
        tensors) and do not have any untracked external dependencies (e.g.,
        perform input/output or access global variables). If you trace such
        models, you will silently get incorrect results on subsequent
        invocations of the model.  You can use `verify=True` to check that the
        original Python code and optimized code are equivalent.
```